RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,568 @@
+
+#### Network upload (Java IDE only)
+
+The Java IDE has an integrated MDNS client that listen for network boards. If a network board is selected the`tools.UPLOAD_RECIPE_ID.upload.network_pattern` is used for upload. `UPLOAD_RECIPE_ID` is obtained in the same way as for the serial upload. A typical recipe is:
+
+`tools.bossac.upload.network_pattern="{runtime.tools.arduinoOTA.path}/bin/arduinoOTA" -address {serial.port} -port 65280 -username arduino -password "{network.password}" -sketch "{build.path}/{build.project_name}.bin" -upload /sketch -b`
+
+please note that:
+
+- the property is called `network_pattern` instead of `pattern`, this is hardcoded
+
+- the port address is still called `{serial.port}` even if it is a network address
+
+- there is a `{network.password}` that is a user entered field, this is hardcoded
+
+### Upload with Pluggable discovery
+
+Given the above, it is clear that the pluggable discovery should provide features to allow:
+
+1. selection of different upload recipes based on the port `protocol`
+1. provide extra port metadata to use on the command line recipe
+1. provide eventual extra info that should be user-supplied (by entering them in a dialog or provided in other ways)
+1. should be backward compatible with the current legacy
+
+#### Recipe selection based on protocol
+
+To allow the above the `UPLOAD_RECIPE_ID` selection is now **dependent on protocol**. The `boards.txt` now must specify a recipe id for each type of upload protocol supported by the board. Previously we used the property `upload.tool=UPLOAD_RECIPE_ID` now this property should be changed to `upload.tool.UPLOAD_PROTOCOL=UPLOAD_RECIPE_ID` so we can choose the correct tool for each protocol. We should also keep the old property for backward compatibility
+
+**Before:**
+
+```
+myboard.upload.tool=bossac
+```
+
+**After:**
+
+```
+# keep for backward compatibility
+myboard.upload.tool=bossac
+# Upload recipes
+myboard.upload.tool.serial=bossac
+myboard.upload.tool.network=arduino_ota
+```
+
+The selected port address will be provided in the variable `{upload.address}`. Other metadata provided by the discover in the `prefs` section will be provided in the `{upload.port.*}` variables.
+
+For backward compatibility we will keep a copy of the address also in `{serial.port}` and in the specific case of a `protocol=serial` we will populate also `{serial.port.file}`.
+
+For example, the following port metadata coming from a pluggable discovery:
+
+```JSON
+{
+  "eventType": "add",
+  "port": {
+    "address": "/dev/ttyACM0",
+    "label": "ttyACM0",
+    "protocol": "serial",
+    "protocolLabel": "Serial Port (USB)",
+    "properties": {
+      "pid": "0x804e",
+      "vid": "0x2341",
+      "serialNumber": "EBEABFD6514D32364E202020FF10181E"
+    },
+    "identificationPrefs": {
+      "pid": "0x804e",
+      "vid": "0x2341"
```

Comment:
As pointed out by @PaulStoffregen, I think this is a leftover from the previous version.
```suggestion
```